### ISSUE TITLE
[dynamo] Refactor switch statements  to improve compile times

### ISSUE THREAD

**Issue by jansel** [pytorch/pytorch#119128]
In many parts of dynamo we have code like:
 ```
if ...

elif ...

elif ...
```
which is `O(n)` (n is number of cases in switch)

We should replace it with:
```
getattr(self, f"method_{name}")(...)
```
or
```
dispatch_table[type(value)](...)
```
which is `O(1)`.

Here is an example of such a change: https://github.com/pytorch/pytorch/issues/119111#issuecomment-1925398694_

And here are some examples of this pattern that still should be fixed:
- [ ] https://github.com/pytorch/pytorch/blob/fcf22a853dc381b5cddd107986a69a2abfc0c8d9/torch/_dynamo/variables/torch.py#L274-L671
- [ ] https://github.com/pytorch/pytorch/blob/7a52455102735a6b79dc688c9b5a1f7bffa8d61f/torch/_dynamo/variables/builder.py#L401-L804
- [x] https://github.com/pytorch/pytorch/blob/7a52455102735a6b79dc688c9b5a1f7bffa8d61f/torch/_dynamo/variables/higher_order_ops.py#L508-L534
- [ ] https://github.com/pytorch/pytorch/blob/7a52455102735a6b79dc688c9b5a1f7bffa8d61f/torch/_dynamo/variables/nn_module.py#L467-L631
- [x] https://github.com/pytorch/pytorch/blob/7a52455102735a6b79dc688c9b5a1f7bffa8d61f/torch/_dynamo/variables/tensor.py#L226-L250
- [ ] https://github.com/pytorch/pytorch/blob/7a52455102735a6b79dc688c9b5a1f7bffa8d61f/torch/_dynamo/variables/lists.py#L238-L293
- [ ] https://github.com/pytorch/pytorch/blob/7a52455102735a6b79dc688c9b5a1f7bffa8d61f/torch/_dynamo/variables/lists.py#L357-L393
- [ ] https://github.com/pytorch/pytorch/blob/7a52455102735a6b79dc688c9b5a1f7bffa8d61f/torch/_dynamo/variables/dicts.py#L210-L275

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @Lucaskabela @oulgen @jamesjwu @aorenste @anijain2305 @laithsakka @masnesral @coconutruben @ezyang @msaroufim @bdhirsh @zou3519 @aakhundov

**Comment by oulgen** [comment]
Thanks, will take a look after pto!

**Comment by oulgen** [comment]
So we have a good way to measure the improvements here or do we just write micro benchmarks and check the nightly benchmarks?

**Comment by ezyang** [comment]
If we're doing micro optimizations like this, it might be useful to use some sort of deterministic performance measure (e.g., as produced by cProfile) which can help us measure very minute differences without having to deal with nondeterminism (at the cost of possibly measuring a distorted metric that doesn't capture true wall time.)

**Comment by jansel** [comment]
I agree setting up some better measurement would be great.  Some of these (that are more on the hot path and run on every op) might show up in nightly benchmarks.  Though many others won't.

**Comment by ragizaki** [comment]
I can start taking a look at the rest of these, implemented the change for `higher_order_ops.py` and would love to take a stab at the rest if possible.

**Comment by ragizaki** [comment]
I can also take a look at setting up some performance measuring, would there by any guidelines or recommendations for that? cc @jansel @ezyang

**Comment by jansel** [comment]
You could pick a benchmark from:

**Comment by ezyang** [comment]
cc @huydhn for reporting these perf numbers

**Comment by anijain2305** [comment]
Our studies have shown that majority of Dynamo time is spent in Fx graph creation and fake tensor prop. Marking this low priority. Contributions are welcome.
